### PR TITLE
Feature/188 generate explicit select list

### LINKED ISSUE #188: Always generate an explicit select list instead of select *
## Problem

There is an optimization on the database side that turns a Virtual Schema SELECT statement containing an explicit list of columns into a SELECT * query when the select list contains all the columns from metadata. 

## Solution

We can use an `involvedTables` field of the pushdown request to create an explicit select list with column names.

## See also

* Fixes exasol/elasticsearch-virtual-schema#4.

### PR DESCRIPTION
**Pull request by AnastasiiaSergienko**
Closes #188.

### CODE REVIEW

File: doc/changes/changes_14.0.0.md
Comment (redcatbear) [inline doc/changes/changes_14.0.0.md:11]:
```suggestion
* #190: Improved `SELECTLIST_EXPRESSIONS` documentation
```

File: src/main/java/com/exasol/adapter/request/parser/PushdownSqlParser.java
Comment (redcatbear) [inline src/main/java/com/exasol/adapter/request/parser/PushdownSqlParser.java:223]: Use new error reporting.
Type: should not happen -> ticket

File: src/main/java/com/exasol/adapter/request/parser/PushdownSqlParser.java
Comment (redcatbear) [inline src/main/java/com/exasol/adapter/request/parser/PushdownSqlParser.java:224]:
```suggestion
                        "Unable to find a metadata for table \"" + table.getName() + "\" during collection of involved columns.");
```

File: src/main/java/com/exasol/adapter/request/parser/PushdownSqlParser.java
Comment (redcatbear) [inline src/main/java/com/exasol/adapter/request/parser/PushdownSqlParser.java:257]:
```suggestion
        final Map<String, TableMetadata> tableMetadataMap = new HashMap<>(this.involvedTablesMetadata.size());
```

File: src/main/java/com/exasol/adapter/request/parser/PushdownSqlParser.java
Comment (redcatbear) [inline src/main/java/com/exasol/adapter/request/parser/PushdownSqlParser.java:693]: Use double quotes. Firstly we are talking about identifiers, secondly to be uniform.

**Review by redcatbear — CHANGES_REQUESTED** [review]

File: src/main/java/com/exasol/adapter/request/parser/PushdownSqlParser.java
Comment (AnastasiiaSergienko) [reply src/main/java/com/exasol/adapter/request/parser/PushdownSqlParser.java:223]: We should do it separately https://github.com/exasol/virtual-schema-common-java/issues/195

File: src/main/java/com/exasol/adapter/request/parser/PushdownSqlParser.java
Comment (AnastasiiaSergienko) [reply src/main/java/com/exasol/adapter/request/parser/PushdownSqlParser.java]: done

**Comment by AnastasiiaSergienko** [comment]
-Exasol dialect integration tests passed.
-Tested manually with the Exasol dialect. 
-Tested manually with MySQL that contains unsupported data types. When I SELECT * from a table with unsupported data types, it returns only supported columns and doesn't throw an exception as it was before.

File: src/main/java/com/exasol/adapter/request/parser/PushdownSqlParser.java
Comment (exanm) [inline src/main/java/com/exasol/adapter/request/parser/PushdownSqlParser.java:693]: There is another `IllegalStateException` in `findInvolvedTableMetadata` without quotes around names.

File: src/main/java/com/exasol/adapter/request/parser/PushdownSqlParser.java
Comment (exanm) [inline src/main/java/com/exasol/adapter/request/parser/PushdownSqlParser.java:202]: I prefer the old if-structure here, because `collectAllInvolvedColumns()` should never return an empty list, so we can return immediately in that case.

Also, if the database unintentionally sends `involvedTables` with a table with an empty column list, the parser never notices this. And then we reach this point with an empty select list. I think the parser should prevent such empty column lists, instead of generating SELECT TRUE here.

**Review by exanm — APPROVED** [review]

**Comment by AnastasiiaSergienko** [comment]
RLS integration tests are fine too. From our side, we're ready to merge the changes

File: src/main/java/com/exasol/adapter/request/parser/PushdownSqlParser.java
Comment (redcatbear) [inline src/main/java/com/exasol/adapter/request/parser/PushdownSqlParser.java:119]: This violates the SoC. You can't have a method called "parse..." that takes an additional parameter that has nothing to do with the part being parsed.
Separate parsing and construction of the select list, please.

This would then look something like this:

```suggestion
        final SqlSelectList originalSelectList = parseOrigianlSelectList(select.getJsonArray("selectList"));
        final SqlSelectList expandedSelectList = createSelectList(originalSelectList, from);
```

**Review by redcatbear — CHANGES_REQUESTED** [review]

File: src/main/java/com/exasol/adapter/request/parser/PushdownSqlParser.java
Comment (AnastasiiaSergienko) [reply src/main/java/com/exasol/adapter/request/parser/PushdownSqlParser.java]: Restructured code

**Review by redcatbear — DISMISSED** [review]

**Comment by chiaradiamarcelo** [comment]
> RLS integration tests are fine too. From our side, we're ready to merge the changes

Waiting for the core team tests to be merged.

File: src/main/java/com/exasol/adapter/request/parser/PushdownSqlParser.java
Comment (exanm) [inline src/main/java/com/exasol/adapter/request/parser/PushdownSqlParser.java:261]: The queue is confusing me. Shouldn't this function return the involved tables in the wrong order, if we have an input similar to the test `pushdown_request_nested_join_without_select_list.json`, but with the join on the left side and the table on the right side? Please add such a test. Somehow the select list in the generated SQL is correct in that case, but I would still prefer a recursive function here instead of a queue, to make this more readable.

Comment (AnastasiiaSergienko) [reply src/main/java/com/exasol/adapter/request/parser/PushdownSqlParser.java:261]: I'm going to test it

**Review by exanm — APPROVED** [review]

**Pull request merged**
